File: compare_embedding_JS_vs_PY/testmodel.py
import os
import json
import shutil
import numpy as np
from deepface import DeepFace


def get_embedding(image_path, model_name="Facenet512"):
    """Get the embedding vector for the given image."""
    embedding = DeepFace.represent(
        img_path=image_path, model_name=model_name, enforce_detection=False
    )
    return embedding[0]["embedding"]


def save_embedding(image_path, output_path):
    """Save the embedding vector to a JSON file."""
    embedding = get_embedding(image_path)
    with open(output_path, "w") as f:
        json.dump(embedding, f)
    print(f"Embedding saved to {output_path}")


def main():
    image_dir = "../imgs/test_imgs"
    output_dir = "./embeddings_py"
    temp_dir = "./temp_images"
    os.makedirs(output_dir, exist_ok=True)
    os.makedirs(temp_dir, exist_ok=True)

    for image_name in os.listdir(image_dir):
        if image_name.endswith(".jpeg"):
            image_path = os.path.join(image_dir, image_name)
            temp_image_path = os.path.join(temp_dir, f"temp_{image_name}")
            shutil.copy(image_path, temp_image_path)
            output_path = os.path.join(output_dir, f"{image_name}.json")
            save_embedding(temp_image_path, output_path)
            os.remove(temp_image_path)


if __name__ == "__main__":
    main()

File: aaa_code_helper/compare_embedding_JS_vs_PY/testmodel.js
const fs = require("fs");
const tf = require("@tensorflow/tfjs-node");
const cv = require("@techstark/opencv-js"); // OpenCV.js for Node.js
const path = require("path");
const sharp = require("sharp");


async function preprocessImage(imagePath) {
  const buffer = await sharp(imagePath).resize(160, 160).toBuffer();
  const tensor = tf.tidy(() => {
    const decodedImage = tf.node.decodeImage(buffer, 3);
    return decodedImage.expandDims(0).toFloat().div(255.0); // Normalized tensor
  });
  return tensor;
}

async function loadModel(modelPath) {
  return await tf.loadGraphModel("file://" + modelPath);
}

async function getEmbedding(model, imagePath) {
  const tensor = await preprocessImage(imagePath);
  const embedding = model.predict(tensor);
  tensor.dispose(); // Free memory
  return embedding.arraySync()[0];
}

async function saveEmbedding(imagePath, outputPath, model) {
  const embedding = await getEmbedding(model, imagePath);
  fs.writeFileSync(outputPath, JSON.stringify(embedding));
  console.log(`Embedding saved to ${outputPath}`);
}

async function main() {
  const imageDir = "../imgs/test_imgs";
  const outputDir = "./embeddings_js";
  const modelPath = path.resolve(__dirname, "./model/FaceNet/Facenet512_tfjs_graph_model/model.json");

  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir);
  }

  const model = await loadModel(modelPath);

  const imageFiles = fs.readdirSync(imageDir).filter(file => file.endsWith(".jpeg"));
  for (const imageName of imageFiles) {
    const imagePath = path.join(imageDir, imageName);
    const outputPath = path.join(outputDir, `${imageName}.json`);
    await saveEmbedding(imagePath, outputPath, model);
  }
}

main().catch(err => console.error(err));
